src/runtime/pkg/containerd-shim-v2/service.go

```diff
@@ -935,6 +935,10 @@ func (s *service) Shutdown(ctx context.Context, r *taskAPI.ShutdownRequest) (_ *
 	s.mu.Lock()
 	if len(s.containers) != 0 {
 		s.mu.Unlock()
+
+		span.End()
+		katatrace.StopTracing(s.rootCtx)
```

Comment:
This is fine, but to avoid the duplicated code it might be better to defer the unlock and the trace ending code. Something like this:

```go
s.mu.Lock()

defer func() {
    s.mu.Lock()
    span.End()
    katatrace.StopTracing(s.rootCtx)
}()

if len(s.containers) != 0 {
    return empty, nil
}

s.cancel()

// ...
```

Reply:
@jodh-intel but since we run `os.Exit()` in `Shutdown`, aren't deferred functions not being run at all? See https://pkg.go.dev/os#Exit

Reply:
@cmaf - Good catch! Yep, unfortunately `defer` isn't an option here, so disregard this suggestion :sob: